File: frontend/src/api/syllables.ts
const API_URL = import.meta.env.VITE_API_URL as string;

export interface SyllableInfo {
  text: string;
  key: string;    // vowel phoneme e.g. "AH"
  stress: number; // 0=unstressed, 1=primary, 2=secondary
}

export interface SyllableOccurrence {
  line: number;
  word_index: number;
  syllable_index: number;
}

export interface SyllableGroup {
  color_index: number;
  phoneme_key: string;
  occurrences: SyllableOccurrence[];
}

export interface SlantOccurrence {
  line: number;
}

export interface SlantGroup {
  color_index: number;
  vowel_key: string;
  occurrences: SlantOccurrence[];
}

export interface AnalyzeResponse {
  line_counts: number[];
  syllable_data: SyllableInfo[][][];  // [line][word][syllable]
  syllable_groups: SyllableGroup[];
  slant_groups?: SlantGroup[];
}

export async function fetchAnalysis(lines: string[]): Promise<AnalyzeResponse> {
  const response = await fetch(`${API_URL}/api/analyze`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ lines }),
  });
  if (!response.ok) throw new Error(`Analyze error: ${response.status}`);
  return response.json() as Promise<AnalyzeResponse>;
}

export async function fetchSyllableCounts(lines: string[]): Promise<number[]> {
  const response = await fetch(`${API_URL}/api/syllables`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ lines }),
  });

  if (!response.ok) {
    throw new Error(`Syllable API error: ${response.status}`);
  }

  const data = (await response.json()) as { counts: number[] };
  return data.counts;
}

File: backend/schemas/syllables.py
from pydantic import BaseModel


class SyllableRequest(BaseModel):
    lines: list[str]


class SyllableResponse(BaseModel):
    counts: list[int]


class AnalyzeRequest(BaseModel):
    lines: list[str]


class SyllableInfo(BaseModel):
    text: str    # e.g. "hun"
    key: str     # vowel phoneme e.g. "AH"; empty if word not in CMU dict
    stress: int = 0  # 0=unstressed, 1=primary, 2=secondary (CMU stress digit)
    onset: str = ""  # space-joined consonant phonemes before the vowel, e.g. "K"
    coda: str = ""   # space-joined consonant phonemes after the vowel, e.g. "T"


class SyllableOccurrence(BaseModel):
    line: int
    word_index: int
    syllable_index: int


class SyllableGroup(BaseModel):
    color_index: int
    phoneme_key: str
    occurrences: list[SyllableOccurrence]


class SlantOccurrence(BaseModel):
    line: int          # which line (0-indexed)
    word_index: int    # which word within the line (0-indexed, whitespace-split)

class SlantGroup(BaseModel):
    color_index: int   # starts after the last syllable_groups color_index
    kind: str           # "assonance" (shared vowel) | "consonance" (shared trailing consonants)
    key: str            # the shared vowel phoneme (assonance) or consonant cluster (consonance)
    occurrences: list[SlantOccurrence]


class AnalyzeResponse(BaseModel):
    line_counts: list[int]
    syllable_data: list[list[list[SyllableInfo]]]  # [line][word][syllable]
    syllable_groups: list[SyllableGroup]
    slant_groups: list[SlantGroup] = []
